### ISSUE TITLE
Auto stop gnirehtet app when disconnected

### ISSUE THREAD

**Issue by twoi** [Genymobile/gnirehtet#51]
In https://github.com/Genymobile/gnirehtet/issues/49, we managed to come up with a script that automatically reestablishes the connection - which is great.

Now I also have a usb wired ethernet device, that I use alternatively with this reverse tethering solution. However, when the gnirehtet app is running on the device - and it stays running after pulling the plug - I cannot use the wired ethernet device until I stop gnirehtet. Then ethernet works, and when I switch back to reverse tethering, that works as well, as #49 restarts the app on the device.

So all I seem to need now is the app to auto-stop when the usb cable is disconnected. Does it make sense to keep it running after unplugging in the first place? Because unplugging and replugging does not do the trick even when the app is running - hence the need for the script in #49 - and #49 does not require the app to stay running either. So unless there are scenarios I am not aware off, the gnirethet app should auto stop by default.

**Comment by Biswa96** [comment]
Yeah, it stops by default when you disconnect USB from PC. The relay server shows client disconnected.

**Comment by twoi** [comment]
@Biswa96 The app on the Android device continues running. The notification item (whatever the technical Android term for this would be) reads "Disconnected from the relay server" and displays a "STOP GNIREHTET" button.

**Comment by rom1v** [comment]
> Does it make sense to keep it running after unplugging in the first place?

The rationale is that a user may use gnirehtet to connect to the _trusted_ desktop network, and don't want to leak data if it switches back automatically to mobile network. That's why I wanted this action to be manual. (In practice, we detect the socket disconnection, not the cable unplug.)

However, we might consider adding an option to get the behavior you request, something like:

    adb start -d   # disable on disconnect

**Comment by twoi** [comment]
That would be perfect!

I haven't looked at the sources yet, and I'm sure you'd be able to implement this feature in 15 minutes, but if you'd prefer a pull request, do let me know.

I guess this would involve changes to all 3 projects:
1) the desktop side java code and
2) the rust code,
to accept and forward the parameter to the android app and
3) the android java app to accept the parameter from the server and actually implement the feature

**Comment by rom1v** [comment]
> Except that -d is already taken - to override the default DNS server.

Ahah, I managed to get a name conflict by only supporting 2 parameters 😆.

> but if you'd prefer a pull request, do let me know

If you're motivated, please do :)

If you want to implement only the java or only the rust part, I'll take care of the other if necessary.

> I guess this would involve changes to all 3 projects

Correct 👍

**Comment by rom1v** [comment]
@twoi Do you still consider implementing the option?

**Comment by svladimirs** [comment]
adb start -d   # disable on disconnect
Where and when I must do this? Before or after '@gnirehtet.exe autorun'?

**Comment by rom1v** [comment]
> Where and when I must do this?

For now, never, it's not implemented.

**Comment by midluk** [comment]
I use gnirehtet when I have wired network on my notebook but only unreliable wifi and/or mobile data connection. I definitely want those connection methods used when I unplug my phone (better have unreliable connection than none at all, and the connection might be better at the place I move my phone to after unplugging). But currently I always have to unlock my phone to stop gnirehtet. There definitely should be an option to stop the app automatically.
I can see the uses of preventing data from going through other untrusted channels, but are you sure that that use case is important for a majority of users?

**Comment by rom1v** [comment]
> There definitely should be an option to stop the app automatically.

I'm ok with that.

> I can see the uses of preventing data from going through other untrusted channels, but are you sure that that use case is important for a majority of users?

I don't know, but IMO it should have the "safe" behavior by default. As an explicit option, it's ok.

**Comment by rom1v** [comment]
I'll probably add an option when I have some time, but meanwhile you can just apply this change:

```diff
diff --git a/app/src/main/java/com/genymobile/gnirehtet/GnirehtetService.java b/app/src/main/java/com/genymobile/gnirehtet/GnirehtetService.java
index 7e8b716..70732cc 100644
--- a/app/src/main/java/com/genymobile/gnirehtet/GnirehtetService.java
+++ b/app/src/main/java/com/genymobile/gnirehtet/GnirehtetService.java
@@ -224,7 +224,7 @@ public class GnirehtetService extends VpnService {
                     break;
                 case RelayTunnelListener.MSG_RELAY_TUNNEL_DISCONNECTED:
                     Log.d(TAG, "Relay tunnel disconnected");
-                    vpnService.notifier.setFailure(true);
+                    vpnService.stop(vpnService);
                     break;
                 default:
             }
```

**Comment by tango25001** [comment]
Can someone tell me, how i get this change into the app? Thank you very much

**Comment by simonferndriger** [comment]
And chance this feature request gets developed at all? Would be awesome if it would! Since reverse thethering is working not in all cases (Google products slowely stop working with it), I need to switch back to wireless internet back and forth currently.

**Comment by tango25001** [comment]
Hello I compiled the app with the above mentioned changes ans it worked fine so far - up to Android 9.

I bought a new Samsung Android 10 device and now the app does not exit automatically. Is there a solution for that?

Thank you!

**Comment by simonferndriger** [comment]
Hi

How can I use this if I can't compile this special feature but want to use it nevertheless?

**Comment by salihmarangoz** [comment]
I have compiled the project with this feature: 
https://github.com/salihmarangoz/gnirehtet/tree/master/COMPILED_BY_ME

If you don't trust the binaries, check the last commit. I did some changes to compile it on Ubuntu 22.04. Tested on Poco X7 Pro and it works with the java version at least. I suppose other versions also work. Just make sure to add "-s" arg to the commandline because that enabled the "stop on disconnect" feature.

Edit: Also added enabled the feature hardcoded for the java version now.

**Comment by simonferndriger** [comment]
Seems to work, thanks!
